### PR TITLE
fix #9276: check valdef is realizable at pickler

### LINKED ISSUE #9276: Val is not stable unless path requested by use site
it seems that a val is not stable in TASTy unless a stable path is requested for it:
## Minimized code
```scala
// before
object Consts {
  val msg = "Hello World" // not stable
}
object TestConsts {
  def foo = Consts.msg
}
```

```scala
// after
object Consts {
  val msg = "Hello World" // now it is stable because of a use site
}
object TestConsts {
  def foo: Consts.msg.type = Consts.msg
}
```

## Output

before:
```scala
80:       VALDEF(7) 26 [msg]
83:         TYPEREF 27 [String]
85:           SHAREDtype 35
87:         STRINGconst 28 [Hello World]
```
after:
```scala
80:       VALDEF(8) 26 [msg]
83:         TYPEREF 27 [String]
85:           SHAREDtype 35
87:         STRINGconst 28 [Hello World]
89:         STABLE
```

### PR DESCRIPTION
**Pull request by bishabosha**
fixes #9276

### CODE REVIEW

**Comment by bishabosha** [comment]
I will repost this comment here:
https://github.com/lampepfl/dotty/issues/9276#issuecomment-804087804

**Comment by smarter** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala [hunk 1]
```diff
@@ -343,6 +343,11 @@ class TreePickler(pickler: TastyPickler) {
         case _ if tpt.isType => pickleTpt(tpt)
       }
       pickleTreeUnlessEmpty(rhs)
+      mdef match
+        case _: ValDef if !sym.is(StableRealizable) =>
```
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala:347]: should probably check the tag instead

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
The problem with this is: checking realizability is usually cheap but has unbounded complexity for lazy vals. We sort of rely on the fact that the intersection of (non-object) lazy vals and vals with type members (where we check stability) is usually small. But with this PR that consideration is now invalid; we will check realizability for every val, at possibly unbounded cost. So, that makes me nervous, even if our benchmarks don't show it.

**Review by odersky — CHANGES_REQUESTED** [review]
I think in summary the potentially unbounded slowdowns weigh more heavily than the potential gains

**Comment by bishabosha** [comment]
close in favor of #11858

**Pull request closed without merging**